### DISCUSSION TITLE
[mini.ai] how to select text between tags without selecting carriage return character(<CR>)

### DISCUSSION THREAD

**Discussion by sjava** [nvim-mini/mini.nvim#271] [General]
Example below: I want to select only 2 lines with \<p> tags when the cursor is somewhere on \<div> or \</div>
```html
<div>
  <p>This is some paragraph.</p>
  <p>And this is other paragraph.</p>
</div>

**Comment by echasnovski** [comment]
You can override [built-in `t` textobject](https://github.com/echasnovski/mini.nvim/blob/main/lua/mini/ai.lua#L1182) by adding `%s*` outside of capture groups. Like so:

```lua
require('mini.ai').setup({
  custom_textobjects = {
    t = { '<(%w-)%f[^<%w][^<>]->.-</%1>', '^<.->%s*().*()%s*</[^/]->$' },
  }
})
```

Then use linewise selection with `Vit`. Depending on your search method, you might need to increase selection again with `it`.

**Reply by sjava** [reply]
That works! Thank you very much!

**Comment by megalithic** [comment]
@echasnovski: as a follow-up to this, is there a way to make vit/cit use nvim-treesitter-textobject's for tag changing/selection/operations instead of of the capture group way from the mini.ai plugin?

the reason being is, with elixir and heex templates you can define custom html tags that ultimately point to an elixir function, so `<div><.custom_tag custom_attr="foo">hi from custom tag</.custom_tag>` would point to a function `def custom_tag(assigns), do: ""` for instance; the present defaults in mini.ai don't seem to work with text object operations on those custom attributes, but the defaults with nvim-treesitter-textobjects do.

here's the impl of the textobject query for the heex filetype: https://github.com/nvim-treesitter/nvim-treesitter-textobjects/blob/master/queries/heex/textobjects.scm

**Reply by echasnovski** [reply]
Yes, it might be possible. Try following suggestions from [this comment](https://github.com/echasnovski/mini.nvim/issues/110#issuecomment-1212277863). In your case the code seems to be exactly the same with `@function.outer` and `@function.inner` for `a` and `i` tag textobjects.

Another solution might be to modify default tag textobject specification to allow punctuation in tag name. It would be something like this (use `[%p%w]` instead of `%w`):

```lua
require('mini.ai').setup({
  custom_textobjects = {
    t = { '<([%p%w]-)%f[^<%p%w][^<>]->.-</%1>', '^<.->().*()</[^/]->$' },
  }
})
```
